proxy/router/src/cache.rs

```diff
@@ -0,0 +1,127 @@
+use indexmap::IndexMap;
+use std::hash::Hash;
+
+// Reexported so IndexMap isn't exposed.
+pub use indexmap::Equivalent;
+
+/// A cache for routes
+///
+/// ## Assumptions
+///
+/// - `access` is common;
+/// - `store` is less common;
+/// - `capacity` is large enough..
+///
+/// ## Complexity
+///
+/// - `access` computes in O(1) time (amortized average).
+/// - `store` computes in O(1) time (average).
+// TODO LRU
+pub struct Cache<K: Hash + Eq, V> {
+    vals: IndexMap<K, V>,
+    capacity: usize,
+}
+
+pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a> {
```

Comment:
Minor nit, but I think it would be nice if there was a doc comment explaining what this is for.